api-select.go

```diff
@@ -0,0 +1,589 @@
+		opts.Type = SelectObjectTypeJSON
+	}
+
+	selectReqBytes, err := xml.Marshal(opts.ToObjectSelectRequest(expression, compressionType))
+
+	if err != nil {
+		return nil, err
+	}
+
+	urlValues := make(url.Values)
+	urlValues.Set("select", "")
+	urlValues.Set("select-type", "2")
+	// Execute POST on bucket/object.
+	resp, err := c.executeMethod(ctx, "POST", requestMetadata{
+		bucketName:       bucketName,
+		objectName:       objectName,
+		queryValues:      urlValues,
+		contentMD5Base64: sumMD5Base64(selectReqBytes),
+		contentSHA256Hex: sum256Hex(selectReqBytes),
+		contentBody:      bytes.NewReader(selectReqBytes),
+		contentLength:    int64(len(selectReqBytes)),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp, bucketName, "")
+	}
+
+	// Process needs to return a eventStream to the client
+	// Wrap this in a Go routine so as to allow the streaming
+	streamer := &Events{
+		Response: make(chan interface{}),
+	}
+	go func() {
+		processBinary(resp.Body, streamer)
+	}()
+	return streamer, nil
+}
+
+// processBinary is the main function that decodes the large byte array into
+// several events that are sent through the eventstream.
+func processBinary(message io.Reader, streamer *Events) {
+	var partialRecordVal []byte
+	lenOfTrim := 0
+	for {
+		var prelude preludeInfo
+		var headers *headerInfo
+		// Create CRC code
+		crc := crc32.New(crc32.IEEETable)
+		crcReader := io.TeeReader(message, crc)
+		var err error
+		var payload []byte
+		// Extract the prelude(12 bytes) into a struct to extract relevant
+		// information.
+		prelude, err = processPrelude(crcReader, crc)
+		if err != nil {
+			if streamer.err == nil {
+				streamer.err = err
+			}
+			continue
+		}
+		// Extract the headers(variable bytes) into a struct to extract relevant
+		// information
+		if prelude.headerLen > 0 {
+			headers, err = extractHeader(io.LimitReader(crcReader, int64(prelude.headerLen)))
+			if err != nil {
+				if streamer.err == nil {
+					streamer.err = err
+				}
+			}
+		}
+		// Calculate the length of the payload so that the appropriate amount of
+		// bytes can be read
+		payloadLen := calcPayloadLen(prelude)
+		// Checks that there is a payload and if there is, then extracts it.
+		if payloadLen > 0 {
+			payload, err = extractPayload(io.LimitReader(crcReader, int64(payloadLen)), payloadLen)
+			// Special case for dealing with messages that are partial.
+			if isPartial(payload) {
+				payload, partialRecordVal, lenOfTrim = processPartial(payload)
+				processEvent(payload, headers, streamer)
+				// return the complete part of the message so far to the client
+				// for loop in case there are multiple sequential partial rows
+				for {
+					// Extracts the partial payload.
+					myPayload, err := extractPayload(io.LimitReader(crcReader, int64(lenOfTrim)), lenOfTrim)
+					if err != nil {
+						if streamer.err == nil {
+							streamer.err = err
```

Comment:
Looks like we should return here closing the events channel - the error could be due to anything (e.g. network error, protocol error).